### DISCUSSION TITLE
pip requirements - are direct references (<package>@<URL> notation) supported?

### DISCUSSION THREAD

**Discussion by chopeen** [renovatebot/renovate#32650] [Request Help] [answered]
### How are you running Renovate?

Self-hosted Renovate

### If you're self-hosting Renovate, tell us which platform (GitHub, GitLab, etc) and which version of Renovate.

GitLab, Docker 39.23-full

### Please tell us more about your question or problem

[PEP440 specification](https://www.python.org/dev/peps/pep-0440/) defines [direct references](https://peps.python.org/pep-0440/#direct-references).

I have the following entry in `requirements.txt` to install a CPU-only PyTorch packages:

```
torch @ https://download.pytorch.org/whl/cpu/torch-2.5.0%2Bcpu-cp311-cp311-linux_x86_64.whl#sha256=217b7de83d1cc71f1de2eae4288cb25a8210a109424a0c1fdde640e3778747d7
```

Will Renovate be able to propose upgrades for a Python dependency defined this way?

**Comment by rarkins** [comment]
You can:

1. Try it and see, or
2. Look in the tests top see if anything similar is there

**Reply by chopeen** [reply, marked as answer]
I guess the answer is "no":

1. Dependency Dashboard does not list `torch` under "Detected dependencies" (all the other dependencies are there) and `torch` does not appear in the logs even when Renovate runs in `DEBUG` mode.
2. Nothing related under `lib/modules/manager/pip_requirements`.
